### PR TITLE
Growing UI nodes Fix

### LINKED ISSUE #8911: Taffy rounding bug can cause UI nodes to expand each update
## Bevy version

main  910f984709fb58cddbb73

## What you did

```rust
use bevy::prelude::*;
use bevy::window::WindowResolution;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins.set(WindowPlugin {
            primary_window: Some(Window {
                resolution: WindowResolution::new(800., 600.).with_scale_factor_override(1.5),
                ..Default::default()
            }),
            ..Default::default()
        }))
        .add_systems(Startup, setup)
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn(Camera2dBundle::default());
    commands
        .spawn(NodeBundle {
            style: Style {
                width: Val::Percent(100.),
                left: Val::Px(1.),
                ..Default::default()
            },
            ..Default::default()
        })
        .with_children(|builder| {
            builder
                .spawn(NodeBundle {
                    style: Style {
                        width: Val::Px(100.),
                        justify_content: JustifyContent::End,
                        ..Default::default()
                    },
                    background_color: Color::BLACK.into(),
                    ..Default::default()
                })
                .with_children(|builder| {
                    builder.spawn(NodeBundle {
                        style: Style {
                            min_width: Val::Px(200.),
                            ..default()
                        },
                        background_color: Color::RED.with_a(0.5).into(),
                        ..Default::default()
                    });
                });
        });
}
```

## What went wrong

https://github.com/bevyengine/bevy/assets/27962798/dd44f597-81a8-4318-99a7-6cb2c195201c

The inner leaf node expands by a physical pixel each time the UI updates.

The problem is with how Taffy rounds the layout coordinates. The rust `round` function rounds away from zero, so x bounds from  -100.5 to 99.5 (width 200) will be rounded to -101 to 100 (width 201), gaining a pixel.

The accumulated error resets after a full update (triggered when the window is resized or the scale factor changes).

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

fixes #8911, #7712

## Solution

Rounding was added to Taffy which fixed issue #7712.
The implementation uses the f32 `round` method which rounds ties (fractional part is a half) away from zero. Issue #8911 occurs when a node's min and max bounds on either axis are "ties" and zero is between them. Then the bounds are rounded away from each other, and the node grows by a pixel. This alone shouldn't cause the node to expand continuously, but I think there is some interaction with the way Taffy recomputes a layout from its cached data that I didn't identify. 

This PR fixes #8911 by first disabling Taffy's internal rounding and using an alternative rounding function that rounds ties up. 
Then, instead of rounding the values of the internal layout tree as Taffy's built-in rounding does, we leave those values unmodified and only the values stored in the components are rounded. This requires walking the tree for the UI node geometry update rather than iterating through a query.

Because the component values are regenerated each update, that should mean that UI updates are idempotent (ish) now and make the growing node behaviour seen in issue #8911 impossible.

I expected a performance regression, but it's an improvement on main:

```
cargo run --profile stress-test --features trace_tracy --example many_buttons
```

ui-rounding-fix-compare

I guess it makes sense to do the rounding together with the node size and position updates.

---

## Changelog

`bevy_ui::layout`:
* Taffy's built-in rounding is disabled and rounding is now performed by `ui_layout_system`.

* Instead of rounding the values of the internal layout tree as Taffy's built-in rounding does, we leave those values unmodified and only the values stored in the components are rounded. This requires walking the tree for the UI node geometry update rather than iterating through a query. Because the component values are regenerated each update, that should mean that UI updates are idempotent now and make the growing node behaviour seen in issue #8911 impossible.

* Added two helper functions `round_ties_up` and `round_layout_coordinates`.

### CODE REVIEW

File: crates/bevy_ui/src/layout/mod.rs [hunk 1]
```diff
@@ -301,33 +304,89 @@ pub fn ui_layout_system(
...
+            &just_children_query,
+            inverse_target_scale_factor as f32,
+            Vec2::ZERO,
+            Vec2::ZERO,
+        );
+    }
+}
+
+#[inline]
+/// Round `value` to the closest whole integer, with ties (values with a fractional part equal to 0.5) rounded towards positive infinity.
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/layout/mod.rs:369]: This doesn't describe or justify the simpler rounding of negative values.

Comment (ickshonpe) [reply crates/bevy_ui/src/layout/mod.rs:369]: I'm not sure what you want me to explain here,  it's just a helper function.

I wasn't sure whether to add it to `bevy_math` instead, but it's probably not that useful outside of this case.

Comment (alice-i-cecile) [reply crates/bevy_ui/src/layout/mod.rs:369]: Reading this code I was surprised by the behavior for negative values being different.

A quick note somewhere about why that is (or even an indication that it's deliberate) would be helpful to reduce reader friction.

I agree that this should remain an internal helper.

Comment (ickshonpe) [reply crates/bevy_ui/src/layout/mod.rs:369]: Ah I get you, I was concentrating too much on the bug fixes that I forgot that this function might be confusing.

I'll add some comments explaining how it works.

**Review by nicoburns — COMMENTED** [review]
This fix looks reasonable to me. I can't reproduce this issue locally in Bevy (example in https://github.com/bevyengine/bevy/issues/8911 doesn't exhibit the growing behaviour for me) so I have been unable to test this, but having looked into https://github.com/DioxusLabs/taffy/issues/502 (which I can reproduce), I suspect this PR might fix the issue even if it was using `.round()` from std rather than `round_ties_up`.

The key change would be the input and output of the rounding process would no longer be the same location. In which case I believe you'd still end up with the width of the node increasing by 1px during rounding, but it would only do so once and then remain that size.

`round_ties_up` would still be a reasonable change to make though.

**Comment by ickshonpe** [comment]
> This fix looks reasonable to me. I can't reproduce this issue locally in Bevy (example in #8911 doesn't exhibit the growing behaviour for me) so I have been unable to test this, 


Oh yes just realised, the example for #8911 is flawed. This was something deliberate that I forgot to fix when I made the PR. I have monitors with different scale factors which makes it trivial to detect UI scaling bugs by moving the window from one screen to the other.

The example adds a left margin:
```
left: UiRect::left(Val::Px(1.)),
```
My main monitor has a scale factor of `1.5`  so `from_style` multiples the margin width by 1.5, giving a 1.5 width so the left side of the inner node has an absolute x position of -148.5. With a scale factor of 1 convert multiplies by 1 and the absolute x position is -99, so there is no half to round away from zero.

> but having looked into [DioxusLabs/taffy#502](https://github.com/DioxusLabs/taffy/issues/502) (which I can reproduce), I suspect this PR might fix the issue even if it was using `.round()` from std rather than `round_ties_up`.

That's right, at first I thought it was only a rounding issue but after I went through the calculations step by step it was clear that the taffy rounding function itself is idempotent and you can see that if you just call `round_layout` multiple times after computing the layout.

> 
> The key change would be the input and output of the rounding process would no longer be the same location. In which case I believe you'd still end up with the width of the node increasing by 1px during rounding, but it would only do so once and then remain that size.
> 
> `round_ties_up` would still be a reasonable change to make though.

Yep, I thought `round_ties_up` might be more expensive as rust's `round` is just an intrinsic but that doesn't seem to be the case, and Taffy's rounding can be disabled anyway. The default should be the most accurate I think, even if the difference is only seen in quite artificial edge cases like this one.

**Comment by mockersf** [comment]
if this bug is fixed in Taffy with https://github.com/DioxusLabs/taffy/issues/501, does it need to also be fixed in Bevy?

**Comment by ickshonpe** [comment]
> if this bug is fixed in Taffy with [DioxusLabs/taffy#501](https://github.com/DioxusLabs/taffy/issues/501), does it need to also be fixed in Bevy?

Even if it is fixed in Taffy before 0.11 is out, it still seems to be a bit more efficient to do the rounding ourselves instead of going over all the layout nodes twice, and we'll probably need to switch to a depth-first walk of the layout tree anyway for multiple UI layout support.

**Review by nicoburns — APPROVED** [review]
This fix looks correct to me. This can and will be fixed upstream in Taffy, but it makes sense to me to fix it here in the meantime.

**Review by cart — APPROVED** [review]
I agree that this is the right call. Just resolved merge conflicts!

**Pull request merged**